pkg.go

```diff
@@ -59,20 +60,19 @@ func (p *Pkg) Resolve(i Importer) {
 		}
 	}
 
-	imports := pkg.Imports
+  //first we set the regular dependencies, then we add the test dependencies
+	//sharing the same set. This allows us to mark all test-only deps linearly
+	unique := make(map[string]struct{})
+	p.setDeps(i, pkg.Imports, pkg.Dir, unique, false)
 	if p.Tree.ResolveTest {
-		imports = append(imports, append(pkg.TestImports, pkg.XTestImports...)...)
+		p.setDeps(i, append(pkg.TestImports, pkg.XTestImports...), pkg.Dir, unique, true)
 	}
-
-	p.setDeps(i, imports, pkg.Dir)
 }
 
 // setDeps takes a slice of import paths and the source directory they are relative to,
 // and creates the Deps of the Pkg. Each dependency is also further resolved prior to being added
 // to the Pkg.
-func (p *Pkg) setDeps(i Importer, imports []string, srcDir string) {
-	unique := make(map[string]struct{})
-
+func (p *Pkg) setDeps(i Importer, imports []string, srcDir string, unique map[string]struct{}, isTest bool) {
```

Comment:
Not to be a contrarian here, but it still is. I just moved it out of the function so that it can be shared between different calls of setDeps. Otherwise, calling setDeps twice with the test imports and regular imports leads to duplicates. That may not be the cleanest solution, however.